drivers/dht/dht.c

```diff
@@ -38,108 +38,185 @@
+    _reset(dev);
 
     xtimer_usleep(2000 * US_PER_MS);
 
     DEBUG("dht_init: success\n");
     return DHT_OK;
 }
 
-int dht_read(const dht_t *dev, int16_t *temp, int16_t *hum)
+int dht_read(dht_t *dev, int16_t *temp, int16_t *hum)
 {
-    uint8_t csum, sum;
+    uint16_t csum, sum;
     uint16_t raw_hum, raw_temp;
 
-    assert(dev && temp && hum);
-
-    /* send init signal to device */
-    gpio_clear(dev->pin);
-    xtimer_usleep(20 * US_PER_MS);
-    gpio_set(dev->pin);
-    xtimer_usleep(40);
-
-    /* sync on device */
-    gpio_init(dev->pin, dev->in_mode);
-    while (!gpio_read(dev->pin)) {}
-    while (gpio_read(dev->pin)) {}
-
-    /*
-     * data is read in sequentially, highest bit first:
-     *  40 .. 24  23   ..   8  7  ..  0
-     * [humidity][temperature][checksum]
-     */
-
-    /* read the humidity, temperature, and checksum bits */
-    raw_hum = read(dev->pin, 16);
-    raw_temp = read(dev->pin, 16);
-    csum = (uint8_t)read(dev->pin, 8);
-
-    /* set pin high again - so we can trigger the next reading by pulling it low
-     * again */
-    gpio_init(dev->pin, GPIO_OUT);
-    gpio_set(dev->pin);
-
-    /* validate the checksum */
-    sum = (raw_temp >> 8) + (raw_temp & 0xff) + (raw_hum >> 8) + (raw_hum & 0xff);
-    if ((sum != csum) || (csum == 0)) {
-        DEBUG("error: checksum invalid\n");
-        return DHT_NOCSUM;
+    assert(dev);
+
+    uint32_t now_us = xtimer_now_usec();
+    if ((now_us - dev->last_read_us) > DATA_HOLD_TIME) {
+        /* send init signal to device */
+        gpio_clear(dev->params.pin);
+        xtimer_usleep(START_LOW_TIME);
+        gpio_set(dev->params.pin);
+        xtimer_usleep(START_HIGH_TIME);
+
+        /* sync on device */
+        gpio_init(dev->params.pin, dev->params.in_mode);
+        if (_wait_for_level(dev->params.pin, 1, TIMEOUT)) {
+            _reset(dev);
+            return DHT_TIMEOUT;
+        }
+
+        if (_wait_for_level(dev->params.pin, 0, TIMEOUT)) {
+            _reset(dev);
+            return DHT_TIMEOUT;
+        }
+
+        /*
+         * data is read in sequentially, highest bit first:
+         *  40 .. 24  23   ..   8  7  ..  0
+         * [humidity][temperature][checksum]
+         */
+
+        /* read the humidity, temperature, and checksum bits */
+        if (_read(&raw_hum, dev->params.pin, 16)) {
+            _reset(dev);
+            return DHT_TIMEOUT;
+        }
+
+        if (_read(&raw_temp, dev->params.pin, 16)) {
+            _reset(dev);
+            return DHT_TIMEOUT;
+        }
+
+        if (_read(&csum, dev->params.pin, 8)) {
+            _reset(dev);
+            return DHT_TIMEOUT;
+        }
+
+        /* Bring device back to defined state - so we can trigger the next reading
+         * by pulling the data pin low again */
+        _reset(dev);
+
+        /* validate the checksum */
+        sum = (raw_temp >> 8) + (raw_temp & 0xff) + (raw_hum >> 8) + (raw_hum & 0xff);
+        if ((sum != csum) || (csum == 0)) {
```

Comment:
The checksum verify won't work this way. According to the datasheet only the lowest 8 bit are used for the checksum. So either add `& 0xFF` or change back the type of `sum` to uint8_t.

Reply:
Also, why is `csum == 0` treated as invalid? Wouldn't that be the correct value for 0.0° C and 0.0% rH?

Reply:
> Also, why is `csum == 0` treated as invalid?

Good question. I didn't author the code, I just changed the indent  level ;-)

> Wouldn't that be the correct value for 0.0° C and 0.0% rH?

According to the data sheet: Yes. And also whenever the sum of the data bytes happen to be zero. I'll fix it.